From 27cefc36506ec052ebedf03430cdc8455bffd3fc Mon Sep 17 00:00:00 2001
From: Keerthy <j-keerthy@ti.com>
Date: Thu, 9 Jan 2020 10:48:55 +0530
Subject: [PATCH 06/25] misc: uclass: Introduce misc_init_by_ofnode

Introduce misc_init_by_ofnode to probe a misc device
using its ofnode.

Signed-off-by: Keerthy <j-keerthy@ti.com>
[Jan: add missing log.h]
Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
---
 drivers/misc/misc-uclass.c | 26 ++++++++++++++++++++++++++
 include/misc.h             |  9 +++++++++
 2 files changed, 35 insertions(+)

diff --git a/drivers/misc/misc-uclass.c b/drivers/misc/misc-uclass.c
index 55381edc98..3b10d6d9ce 100644
--- a/drivers/misc/misc-uclass.c
+++ b/drivers/misc/misc-uclass.c
@@ -5,7 +5,10 @@
 
 #include <common.h>
 #include <dm.h>
+#include <dm/device-internal.h>
+#include <dm/uclass-internal.h>
 #include <errno.h>
+#include <log.h>
 #include <misc.h>
 
 /*
@@ -65,6 +68,29 @@ int misc_set_enabled(struct udevice *dev, bool val)
 	return ops->set_enabled(dev, val);
 }
 
+int misc_init_by_ofnode(ofnode node)
+{
+	struct udevice *dev = NULL;
+	struct uclass *uc;
+	int ret;
+
+	ret = uclass_get(UCLASS_MISC, &uc);
+	if (ret)
+		return ret;
+
+	uclass_foreach_dev(dev, uc) {
+		if (ofnode_equal(node, dev_ofnode(dev))) {
+			ret = device_probe(dev);
+			if (ret)
+				debug("%s: Failed to initialize - %d\n",
+				      dev->name, ret);
+			return ret;
+		}
+	}
+
+	return -ENODEV;
+}
+
 UCLASS_DRIVER(misc) = {
 	.id		= UCLASS_MISC,
 	.name		= "misc",
diff --git a/include/misc.h b/include/misc.h
index 12d1325ee2..79263ed480 100644
--- a/include/misc.h
+++ b/include/misc.h
@@ -76,6 +76,15 @@ int misc_call(struct udevice *dev, int msgid, void *tx_msg, int tx_size,
  */
 int misc_set_enabled(struct udevice *dev, bool val);
 
+/**
+ * misc_init_by_ofnode() - Probe a misc device by using ofnode.
+ * @node: ofnode of the misc device.
+ *
+ * A misc device is probed using ofnode.
+ *
+ * Return: -ve on error, 0 on success
+ */
+int misc_init_by_ofnode(ofnode node);
 /*
  * struct misc_ops - Driver model Misc operations
  *
-- 
2.17.1

